### ISSUE TITLE
vembrane table gives error with multi-sample VCF missing FORMAT entries

### ISSUE THREAD

**Issue by sci-kai** [vembrane/vembrane#171]
We are using vembrane on a multi-sample VCF files and got the following error message:

```
vembrane only supports records with one alternative allele.
Please split multi-allelic records first, for example with `bcftools norm -m-any […]` or `gatk LeftAlignAndTrimVariants […] --split-multi-allelics` or `vcfmulti2oneallele […]`
```

The command looks like this:
```
vembrane table \
    --output test.tsv \
    --header 'CHROM,POS,ID,REF,ALT,QUAL,FILTER,for_each_sample(lambda sample: f"allele_depth_{sample}")' \
    --annotation-key CSQ \
    'CHROM,POS,ID,REF,ALT,QUAL,FILTER,for_each_sample(lambda s: FORMAT["AD"][s])' \
    test.vcf
```

and an entry in the VCF file that produces this error looks like this:

```
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	tumor	normal
chr17	7673803	.	G	A	.	.	AS_SB_TABLE=15,152|5,70;DP=249;ECNT=1;MBQ=36,34;MFRL=289,248;MMQ=60,60;MPOS=26;NALOD=2.20;NLOD=46.46;POPAF=6.00;TLOD=270.33	GT:AD:AF:DP:F1R2:F2R1:FAD:SB	0/1:11,75:0.864:86:6,37:5,36:11,75:1,10,5,70	./.:.:.:.:.:.:.:.
```

The entries are without an alternative allele and normalized using `bcftools norm`.
After some trial & error, I figured the error message only occurs if two things collide:
1. Including the "for_each_sample" expression with a FORMAT field with multiple entries, e.g. "AD" (works with "DP")
2. At least one sample does not has entries in the FORMAT fields (as the normal sample in the example)

Ideally, the output of this TSV file should be NA for missing entry in the FORMAT field for the normal sample.
Can you help me with this problem?

**Comment by tedil** [comment]
Hm, I cannot reproduce the error you are getting using the following minimal example:
```
##fileformat=VCFv4.3
##contig=<ID=chr17,length=84276897>
##FILTER=<ID=PASS,Description="All filters passed">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=AD,Number=.,Type=Integer,Description="…">
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	tumor	normal
chr17	7673803	.	G	A	.	.	.	GT:AD	0/1:11,75	./.:.
```
and either of
```sh
vembrane table --header 'CHROM,ALT,for_each_sample(lambda sample: f"allele_depth_{sample}")' "CHROM, ALT, for_each_sample(lambda sample: (FORMAT['AD'].get(sample, None)))" test.vcf
```
or
```sh
vembrane table --long --header "CHROM,ALT,AD" "CHROM, ALT, FORMAT['AD'][SAMPLE]" test.vcf
```
neither with the current release 1.0.3 nor the current development state.
I suspect that either or both of FORMAT/AD and INFO/AS_SB_TABLE are defined with `Number=A` (or `Number=R`) in the header and bcftools norm did for some reason not modify these.


So, given the following file
```
##fileformat=VCFv4.3
##contig=<ID=chr17,length=84276897>
##FILTER=<ID=PASS,Description="All filters passed">
##INFO=<ID=FOO,Number=A,Type=Integer,Description="Bar">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=AD,Number=A,Type=Integer,Description="…">
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	tumor	normal
chr17	7673803	.	G	A,C	.	.	FOO=1,2	GT:AD	0/1:11,75	./.:.
```
running `bcftools norm -m-any` gets you
```
##fileformat=VCFv4.3
##FILTER=<ID=PASS,Description="All filters passed">
##contig=<ID=chr17,length=84276897>
##INFO=<ID=FOO,Number=A,Type=Integer,Description="Bar">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=AD,Number=A,Type=Integer,Description="…">
##bcftools_normVersion=1.17+htslib-1.17
##bcftools_normCommand=norm -m-any test_multi.vcf; Date=Sun Jul  7 21:18:07 2024
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	tumor	normal
chr17	7673803	.	G	A	.	.	FOO=1	GT:AD	0/1:11	./.:.
chr17	7673803	.	G	C	.	.	FOO=2	GT:AD	0/0:75	./.:.
```
… and this is fine for vembrane, as -- even though the header says `Number=A` -- there is only one value associated with the respective keys `FOO` and `AD`.
Could you provide a bit more of the header? At least the definitions of the keys that are used in the expression?

(Also, if you want to control what a missing value should be replaced with, you can always use `FORMAT['foo'].get(sample, some_default)` instead of the default `NA`)

**Comment by sci-kai** [comment]
Hi, thanks for the quick reply.
Here is the header of the respective VCF file producing the error:

```
##fileformat=VCFv4.2
##FILTER=<ID=PASS,Description="All filters passed">
##FORMAT=<ID=AD,Number=R,Type=Integer,Description="Allelic depths for the ref and alt alleles in the order listed">
##FORMAT=<ID=AF,Number=A,Type=Float,Description="Allele fractions of alternate alleles in the tumor">
##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Approximate read depth (reads with MQ=255 or with bad mates are filtered)">
##FORMAT=<ID=F1R2,Number=R,Type=Integer,Description="Count of reads in F1R2 pair orientation supporting each allele">
##FORMAT=<ID=F2R1,Number=R,Type=Integer,Description="Count of reads in F2R1 pair orientation supporting each allele">
##FORMAT=<ID=FAD,Number=R,Type=Integer,Description="Count of fragments supporting each allele.">
##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Genotype Quality">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=PGT,Number=1,Type=String,Description="Physical phasing haplotype information, describing how the alternate alleles are phased in relation to one another; will always be heterozygous and is not intended to describe called alleles">
##FORMAT=<ID=PID,Number=1,Type=String,Description="Physical phasing ID information, where each unique ID within a given sample (but not across samples) connects records within a phasing group">
##FORMAT=<ID=PL,Number=G,Type=Integer,Description="Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification">
##FORMAT=<ID=PS,Number=1,Type=Integer,Description="Phasing set (typically the position of the first variant in the set)">
##FORMAT=<ID=SB,Number=4,Type=Integer,Description="Per-sample component statistics which comprise the Fisher's Exact Test to detect strand bias.">
##INFO=<ID=AS_SB_TABLE,Number=1,Type=String,Description="Allele-specific forward/reverse read counts for strand bias tests. Includes the reference and alleles separated by |.">
##INFO=<ID=AS_UNIQ_ALT_READ_COUNT,Number=A,Type=Integer,Description="Number of reads with unique start and mate end positions for each alt at a variant site">
##INFO=<ID=CONTQ,Number=1,Type=Float,Description="Phred-scaled qualities that alt allele are not due to contamination">
##INFO=<ID=DP,Number=1,Type=Integer,Description="Approximate read depth; some reads may have been filtered">
##INFO=<ID=ECNT,Number=1,Type=Integer,Description="Number of events in this haplotype">
##INFO=<ID=GERMQ,Number=1,Type=Integer,Description="Phred-scaled quality that alt alleles are not germline variants">
##INFO=<ID=MBQ,Number=R,Type=Integer,Description="median base quality by allele">
##INFO=<ID=MFRL,Number=R,Type=Integer,Description="median fragment length by allele">
##INFO=<ID=MMQ,Number=R,Type=Integer,Description="median mapping quality by allele">
##INFO=<ID=MPOS,Number=A,Type=Integer,Description="median distance from end of read">
##INFO=<ID=NALOD,Number=A,Type=Float,Description="Negative log 10 odds of artifact in normal with same allele fraction as tumor">
##INFO=<ID=NCount,Number=1,Type=Integer,Description="Count of N bases in the pileup">
##INFO=<ID=NLOD,Number=A,Type=Float,Description="Normal log 10 likelihood ratio of diploid het or hom alt genotypes">
##INFO=<ID=OCM,Number=1,Type=Integer,Description="Number of alt reads whose original alignment doesn't match the current contig.">
##INFO=<ID=PON,Number=0,Type=Flag,Description="site found in panel of normals">
##INFO=<ID=POPAF,Number=A,Type=Float,Description="negative log 10 population allele frequencies of alt alleles">
##INFO=<ID=ROQ,Number=1,Type=Float,Description="Phred-scaled qualities that alt allele are not due to read orientation artifact">
##INFO=<ID=RPA,Number=R,Type=Integer,Description="Number of times tandem repeat unit is repeated, for each allele (including reference)">
##INFO=<ID=RU,Number=1,Type=String,Description="Tandem repeat unit (bases)">
##INFO=<ID=SEQQ,Number=1,Type=Integer,Description="Phred-scaled quality that alt alleles are not sequencing errors">
##INFO=<ID=STR,Number=0,Type=Flag,Description="Variant is a short tandem repeat">
##INFO=<ID=STRANDQ,Number=1,Type=Integer,Description="Phred-scaled quality of strand bias artifact">
##INFO=<ID=STRQ,Number=1,Type=Integer,Description="Phred-scaled quality that alt alleles in STRs are not polymerase slippage errors">
##INFO=<ID=TLOD,Number=A,Type=Float,Description="Log 10 likelihood ratio score of variant existing versus not existing">
##MutectVersion=2.2
##contig=<ID=chr1,length=248956422>
##contig=<ID=chr2,length=242193529>
##contig=<ID=chr3,length=198295559>
##contig=<ID=chr4,length=190214555>
##contig=<ID=chr5,length=181538259>
##contig=<ID=chr6,length=170805979>
##contig=<ID=chr7,length=159345973>
##contig=<ID=chr8,length=145138636>
##contig=<ID=chr9,length=138394717>
##contig=<ID=chr10,length=133797422>
##contig=<ID=chr11,length=135086622>
##contig=<ID=chr12,length=133275309>
##contig=<ID=chr13,length=114364328>
##contig=<ID=chr14,length=107043718>
##contig=<ID=chr15,length=101991189>
##contig=<ID=chr16,length=90338345>
##contig=<ID=chr17,length=83257441>
##contig=<ID=chr18,length=80373285>
##contig=<ID=chr19,length=58617616>
##contig=<ID=chr20,length=64444167>
##contig=<ID=chr21,length=46709983>
##contig=<ID=chr22,length=50818468>
##contig=<ID=chrX,length=156040895>
##contig=<ID=chrY,length=57227415>
##contig=<ID=chrM,length=16569>
```

**Comment by sci-kai** [comment]
I figured out that the problem is with missing entries in this case in the AD FORMAT field.
Here is a minimal example showing the issue:

vembrane command (or any of your commands above)

```
vembrane table \
    --output test.tsv \
    --header 'for_each_sample(lambda sample: f"allele_depth_{sample}")' \
    'for_each_sample(lambda s: FORMAT["AD"][s])' \
    test.vcf
```

test.vcf file:

```
##fileformat=VCFv4.2
##FORMAT=<ID=AD,Number=R,Type=Integer>
##contig=<ID=chr17,length=83257441>
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	tumor	normal
chr17	7673803	.	G	A	.	.	.	AD	11,75	.
```

produces this error:

```
vembrane only supports records with one alternative allele.
Please split multi-allelic records first, for example with `bcftools norm -m-any […]` or `gatk LeftAlignAndTrimVariants […] --split-multi-allelics` or `vcfmulti2oneallele […]`
```

The error is resolved in two cases:
1) If either the header of AD is changed to `##FORMAT=<ID=AD,Number=.,Type=Integer>` producing the output
```
allele_depth_tumor      allele_depth_normal
(11, 75)        (None,)
```

2) The respective VCF entry is changed to `chr17	7673803	.	G	A	.	.	.	AD	11,75	.,.`  producing the output
```
allele_depth_tumor      allele_depth_normal
(11, 75)        (None,None)
```

**Comment by sci-kai** [comment]
Thanks, this is now resolving the error.
Regarding the point with the difference between a single missing value and multiple missing entries:
For the usage , e.g. with `vembrane table`, it is important that also accessing the different alleles, e.g., with FORMAT["AD"][sample][1] should produce a single missing value in the TSV file in both cases.

**Comment by tedil** [comment]
Okay, for Number=R this is possible, since that indicates there should be 2 values in the tuple, not only 1. I've pushed a corresponding fix.

**Comment by sci-kai** [comment]
Thanks, I tried the fix and this is working now.
Can you push another release with this fix, so I can incorporate it in our pipeline?

**Comment by tedil** [comment]
Fixed in v1.0.6
